### ISSUE TITLE
Splitting analysis across multiple files - necessity of source() each time?

### ISSUE THREAD

**Issue by francisbarton** [workflowr/workflowr#197]
I think I know what the answer to this is, but I just want to check that I'm not missing a trick or a feature of workflowr.

So I started off just with `analysis/index.Rmd`, and created a whole load of code in `code/` to support the analysis. My `index.Rmd` starts off by using `source()` to bring in a file that I've called `code/00_setup.R` (which in turn sources and runs a load of other code files). Now my index.Rmd has got rather long, so I want to split it into different pages on the resulting site. But it seems that each new `.Rmd` file needs to re-`source()` all the data.

It's my first time using workflowr and I assumed that once `wflow_build()` had worked its way through `index.Rmd`, it would have attached all the necessary libraries, and created all the data objects and functions in the environment that the other Rmd files would need, so they wouldn't each need to load them all over again. My expectation was that the other files would be in a way _subsidiary_ to index.Rmd, and flow from it. But it seems that they are quite independent in terms of their code environment.

So my question is, do I need to create a new `source()` call at the top of each of my site's sub-pages? [It seems that I do](https://jdblischak.github.io/workflowr/articles/wflow-07-common-code.html) (obviously I am getting errors at the moment when I try to build). But am I missing a shortcut that would help here? Something that would tell workflowr to build e.g. `page2.Rmd` using the same environment that was already created for `index.Rmd`?

There's no great cost to me re-`source()`-ing on every page, it just feels a bit clumsy.
Thanks for your advice.

edit: I just realised another possible part of this issue: I had assumed I could set some params for the whole site in the YAML header of `index.Rmd` (Single Source of Truth) and they would be inherited across the other pages, but now I'm wondering if I might need to re-iterate the params on every single page. That would be a pain. I haven't tested this yet though.

(note: possible duplicate of #9  and #164 - I'm not sure)

**Comment by francisbarton** [comment]
I suspect this would be solved if I could force `wflow_build()` to build `index.Rmd` first - then I don't really care what order the other files get built.

**Comment by jdblischak** [comment]
@francisbarton You are correct that each Rmd file in the workflowr project is run in an isolated environment. Thus if you have functions in `code/` that you want to use in each of the Rmd files, then `source()` is likely your best option.

If you need one file to be run prior to all the rest, I recommend using prefixes since `wflow_build()` uses alphanumeric sorting, e.g. `01-setup.Rmd`, `02-clean.Rmd`, `03-model.Rmd`, etc. I don't recommend putting critical code in`index.Rmd`. Because that file is required for rmarkdown websites (which is what workflowr extends), you can't rename it.

Note that the behavior of `wflow_build()` is a little tricky since it has so many options. The default when no files are specified is to build the outdated files in alphanumeric order. However, if you list a file, by default only it is built. And if you list a file and set `make = TRUE`, then the explicitly listed files will be run first followed by the remaining sorted files.

In other words, in your current setup, you could ensure that `index.Rmd` was always built first by running the below:

```
wflow_build("analysis/index.Rmd", make = TRUE)
```

Truly improved support for your issue would require file dependency support like that described in #9. The problem is that workflowr already has a lot of moving parts, so implementing this correctly will be non-trivial.

**Comment by francisbarton** [comment]
Thank you John for that really helpful and swift response!

- your idea of using prefixes is a good one - I might try that but I suspect it won't help in my particular use case
- I think this issue really boils down to me making an incorrect assumption about what would happen when I split my index.Rmd up into smaller files. There's nothing wrong with the workflowr documentation, I just misunderstood what kind of package it is. I'll probably just merge all my pages back in to a single file (at the cost of losing paginated navigation via the top menu)
- I could also solve my issue by repeating `source("code/00_setup.R")` at the top of each file, or by sourcing a custom file for each page.

On my point about sharing params across the whole project, I might be able to do this by specifying them in _site.yml rather than at the top of index.Rmd
